target/ppc/translate/vsx-impl.c.inc

```diff
@@ -2140,80 +2139,54 @@ static bool do_lstxv(DisasContext *ctx, arg_D *a, bool store, bool paired)
 
     ea = tcg_temp_new();
 
-    do_ea_calc(ctx, a->ra, tcg_constant_tl(a->si), ea);
+    mop = MO_Q ^ ctx->default_tcg_memop_mask;
+    le = ctx->le_mode;
+
+    do_ea_calc(ctx, a->ra, tcg_constant_tl(a->si + ((paired && le)*16)), ea);
 
     gen_set_access_type(ctx, ACCESS_INT);
 
-    xt1 = tcg_temp_new_i64();
-    xt2 = tcg_temp_new_i64();
-    xt3 = tcg_temp_new_i64();
-    xt4 = tcg_temp_new_i64();
+    xt = tcg_temp_new_i64();
+
+    /*
+     * If big-endian vsr64_offset(n, !le) returns vsx register first 64bits,
+     * otherwise returns last 64bits.
+     * If big-endian vsr64_offset(n, le) returns vsx register last 64bits,
+     * otherwise returns first 64bits.
+     */
 
     if(store) {
-        get_cpu_vsrh(xt1, a->rt);
-        get_cpu_vsrl(xt2, a->rt);
-        if (ctx->le_mode) {
-            if (paired) {
-                get_cpu_vsrh(xt3, a->rt+1);
-                get_cpu_vsrl(xt4, a->rt+1);
-                tcg_gen_qemu_st_i64(xt4, ea, ctx->mem_idx, MO_LEQ);
-                tcg_gen_addi_tl(ea, ea, 8);
-                tcg_gen_qemu_st_i64(xt3, ea, ctx->mem_idx, MO_LEQ);
-                tcg_gen_addi_tl(ea, ea, 8);
-            }
-            tcg_gen_qemu_st_i64(xt2, ea, ctx->mem_idx, MO_LEQ);
-            tcg_gen_addi_tl(ea, ea, 8);
-            tcg_gen_qemu_st_i64(xt1, ea, ctx->mem_idx, MO_LEQ);
-        } else {
-            tcg_gen_qemu_st_i64(xt1, ea, ctx->mem_idx, MO_BEQ);
+        tcg_gen_ld_i64(xt, cpu_env, vsr64_offset(a->rt, !le));
```

Comment:
What about (not tested):

```
target_long offset;

if (le) {
    do_ea_calc(ctx, a->ra, tcg_constant_tl(a->si + (paired ? 24 : 8)), ea);
    offset = -8;
} else {
    do_ea_calc(ctx, a->ra, tcg_constant_tl(a->si), ea);
    offset = 8;
}
```
Then, for both loads and stores, something like:
```
<load_or_store_code> for ea <-> cpu_vsrh(a->rt) with mop set accordingly
tcg_gen_addi_tl(ea, ea, offset);
<load_or_store_code> for ea <-> cpu_vsrl(a->rt) with mop set accordingly
if (paired) {
    tcg_gen_addi_tl(ea, ea, offset);
    <load_or_store_code> for ea <-> cpu_vsrh(a->rt + 1) with mop set accordingly
    tcg_gen_addi_tl(ea, ea, offset);
    <load_or_store_code> for ea <-> cpu_vsrl(a->rt + 1) with mop set accordingly
}

```

Reply:
If we don't even use 'le' again later, we can just use ``if (ctx->le_mode)`` and drop 'le'.